### DISCUSSION TITLE
Specifying button width with built-in CSS classes

### DISCUSSION THREAD

**Discussion by vedranmiletic** [squidfunk/mkdocs-material#3927] [Q&A] [answered]
### Contribution guidelines

- [X] I've read the [contribution guidelines](https://github.com/squidfunk/mkdocs-material/blob/master/CONTRIBUTING.md) and wholeheartedly agree

### I want to suggest an idea and checked that ...

- [X] ... to my best knowledge, my idea wouldn't break something for other users
- [X] ... the documentation does not mention anything about my idea
- [X] ... there are no open or closed issues that are related to my idea

### Description

[Buttons](https://squidfunk.github.io/mkdocs-material/reference/buttons/) section in the documentation describes the usage of `.md-button` and `.md-button--primary` CSS classes. It would be nice if there were additional CSS classes documented in the same place that could be used to make buttons 100% wide on small screens. Ideally, there would be CSS classes for specifying other widths (1/2, 1/3, 1/4 etc.) that a user could specify for different screen widths.

### Use Cases

E.g. making the buttons equally wide for eye candy.

**Comment by squidfunk** [comment]
Thanks for suggesting! However, this is a prime example of [customization](https://squidfunk.github.io/mkdocs-material/customization/) via [additional CSS](https://squidfunk.github.io/mkdocs-material/customization/#additional-css). I think it is very important that we don't let the Material for MkDocs documentation turn into a general web development guide, so I'd rather recommend asking pure CSS-directed questions like this on [StackOverflow](https://stackoverflow.com/) or on our [discussions board](https://github.com/squidfunk/mkdocs-material/discussions). After all, this has nothing to do with this project in particular. We need to keep our focus sharp.

Converting to discussion.

**Comment by vedranmiletic** [comment]
Fine. Does Material for MkDocs provide ready-made classes for this in a way e.g. Tailwind CSS or Bootstrap does?

**Reply by squidfunk** [reply, marked as answer]
Material for MkDocs is written entirely from scratch without the help of such a framework, so no, there are no such classes. However, you might be able to use helper classes from one of those frameworks, adding it via additional CSS. Learning some CSS might also be an idea if you want to customize the theme. Maybe this is a start (untested):

``` css
@media (max-width: 768px) {
  .md-button--stretch {
    width: 100%;
  }
}
```

Then add:

``` markdown
[Link](#){ .md-button.md-button--stretch }
```

You have to understand that, in general, I can't help users with customizing their installations, because if I would agree to do that, this project won't move forward because I will be doing nothing else except helping everybody achieve precisely what he/she wants. This only benefits the individual and not the whole project. Thus, I can only point you in the right direction. I hope for your understanding.

**Reply by feasgal** [reply]
Thanks! I needed to be able to arrange buttons in a grid that looked uniform, and this CSS:

`.md-typeset .md-button--stretch {
  width: 100%;
}`

in combination with this markdown:

`[Log-in Problem](#){ .md-button .md-button--stretch }`

did the exact job I needed:

<img width="708" alt="image" src="https://user-images.githubusercontent.com/97999520/193948341-b41f8639-7953-4f25-a112-5f4993d116fb.png">

**Reply by Sanrax** [reply]
> <img alt="image" width="708" src="https://user-images.githubusercontent.com/97999520/193948341-b41f8639-7953-4f25-a112-5f4993d116fb.png">

How did you get "Different Account Issue" to be inline with "Log-in Problem"? When I tried using this css and markdown, it makes both buttons 100% width, but they are stacked on top of each other.

**Reply by feasgal** [reply]
They're in a [grid](https://squidfunk.github.io/mkdocs-material/reference/grids/).

```
<div class="grid" markdown>

[Log-in Problem](#){ .md-button .md-button--stretch }

[Different Account Issue](#){ .md-button .md-button--stretch }

[Problem with My Device](#){ .md-button .md-button--stretch }

[Help with an Order](#){ .md-button .md-button--stretch }

[Failed License Move](#){ .md-button .md-button--stretch }

[Different Issue](#){ .md-button .md-button--stretch }

</div>
```

**Reply by Sanrax** [reply]
I see, thank you!

**Reply by Sanrax** [reply]
> They're in a [grid](https://squidfunk.github.io/mkdocs-material/reference/grids/).
> 
> ```
> <div class="grid" markdown>
> 
> [Log-in Problem](#){ .md-button .md-button--stretch }
> 
> [Different Account Issue](#){ .md-button .md-button--stretch }
> 
> [Problem with My Device](#){ .md-button .md-button--stretch }
> 
> [Help with an Order](#){ .md-button .md-button--stretch }
> 
> [Failed License Move](#){ .md-button .md-button--stretch }
> 
> [Different Issue](#){ .md-button .md-button--stretch }
> 
> </div>
> ```

I was only able to get back to my project and start working on it again now, but I just realized that this is a sponsor-only feature, is that right? Would there be any way to replicate this in the free version? I really don't want to have to pay for a hobby project.

**Reply by feasgal** [reply]
Ah. Yes, both grids and card grids are still sponsors-only as of this writing. So without giving away squidfunk's IP, I'll suggest that you could [use additional CSS](https://squidfunk.github.io/mkdocs-material/customization/#additional-css) to define a `my_awesome_grid` class, set its `display` property to `grid`, and play with the [other grid-related properties](https://www.w3schools.com/css/css_grid.asp) until it looks the way you want it to.

```
.md-typeset .my_awesome_grid {
  <properties here>
}
```

**Reply by Sanrax** [reply]
Thank you for the advice, I'll try this out right now!

**Reply by Sanrax** [reply]
> Ah. Yes, both grids and card grids are still sponsors-only as of this writing. So without giving away squidfunk's IP, I'll suggest that you could [use additional CSS](https://squidfunk.github.io/mkdocs-material/customization/#additional-css) to define a `my_awesome_grid` class, set its `display` property to `grid`, and play with the [other grid-related properties](https://www.w3schools.com/css/css_grid.asp) until it looks the way you want it to.
> 
> ```
> .md-typeset .my_awesome_grid {
>   <properties here>
> }
> ```

Well, I've gotten it to work... mostly. My only remaining issue is there is a big gap between the first row and second row of buttons, which I simply can't manage to get rid of, and I'm not sure what is causing it.

Here's my extra.css and markdown:
![image](https://github.com/squidfunk/mkdocs-material/assets/61475057/b3ea8258-d3bc-43a3-8411-64c05c622143)

And the result:
![image](https://github.com/squidfunk/mkdocs-material/assets/61475057/d1a01ce0-57d2-47eb-bc01-b9bc71f39d43)

Would you happen to have any ideas?

**Reply by feasgal** [reply]
I'm suspicious of `grid-auto-rows: 1fr;`. That's making all the rows in the grid be the same height, and it will be based on the length of the content in that first button in the second row. Maybe set `grid-auto-rows: auto;` (or just remove the property; `auto` is its default value) and then set your `row-gap` the same as your `column-gap` (or just use `grid-gap` to set them both).

**Reply by Sanrax** [reply]
> I'm suspicious of `grid-auto-rows: 1fr;`. That's making all the rows in the grid be the same height, and it will be based on the length of the content in that first button in the second row. Maybe set `grid-auto-rows: auto;` (or just remove the property; `auto` is its default value) and then set your `row-gap` the same as your `column-gap` (or just use `grid-gap` to set them both).

I know that's what it does, I want all the boxes to be the same height irregardless of their contents. So I set that to 1fr.

The big gap between first row and second row is there even if I do a very basic grid without the extra lines. It'll just mean that the boxes will be different heights, but the gap is still there.

The only way to fix it that I found so far is to remove the spaces in between each button line in the markdown document, like this:

`[Button1](#){.mdbutton .mdbutton--stretch}`
`[Button2](#){.mdbutton .mdbutton--stretch}`
`[Button3](#){.mdbutton .mdbutton--stretch}`
`[Button4](#){.mdbutton .mdbutton--stretch}`

This however causes another issue- the made with mkdocs footer and the rest of the page bugs out and displays under some parts of the button grid when I resize the screen to a vertical strip like a phone UI.

If I add spaces in between each markdown line, then it renders without bugs, but has that big gap between the first and second line. Pretty confusing issue honestly.

**Comment by vedranmiletic** [comment]
Sure, thanks for your help!
